### ISSUE TITLE
provide a hook to stop training at any point

### ISSUE THREAD

**Issue by david-alexander-white** [PyTorchLightning/pytorch-lightning#1406]
## 🚀 Feature

I want to be able to stop the trainer at any point based on arbitrary logic.
provide a hook to stop training at any point -- not just epoch end, and not just when a metric stop improving

### Motivation

The current EarlyStoppingCallback method only lets you stop at end of epoch, and that only when a metric stops improving by some amount. Other than that the only way I can see to stop training is to throw an exception and then call training teardown.

See https://pytorch-lightning.slack.com/archives/CQXV8BRH9/p1586281898171700

### Pitch

I want a function in the trainer like "end training" that I can call from anywhere in lightning which will cause training to stop, say, at the end of the current batch. 

### Alternatives

Throw an exception? Force the batch to return -1?

**Comment by williamFalcon** [comment]
use a callback for that. and end training ```on_batch_end```

**Comment by williamFalcon** [comment]
https://pytorch-lightning.readthedocs.io/en/0.7.1/callbacks.html

**Comment by david-alexander-white** [comment]
> https://pytorch-lightning.readthedocs.io/en/0.7.1/callbacks.html

thanks for the quick reply!

however... what should I do inside my callback? As far as I can tell there is no way to get training to actually stop. The way EarlyStoppingCallback does it is actually hard-coded into the training loop, in a way we couldn't do on on_batch_end

**Comment by david-alexander-white** [comment]
simplest way to fix it i think would be to add a binary flag to the trainer that's like "stop_requested" or something like that, and u check it after every batch, and if it's on, you return from train_epoch, and you check it again after each epoch

**Comment by david-alexander-white** [comment]
kind of duct tape but i'm happy to submit a PR

**Comment by williamFalcon** [comment]
@PyTorchLightning/core-contributors

**Comment by david-alexander-white** [comment]
oh, the other way would be to just throw a custom exception. that's what's done for ctrl-c currently and what i'll be using for my own use case for now

**Comment by awaelchli** [comment]
The docs for [training_step ](https://pytorch-lightning.readthedocs.io/en/latest/pytorch_lightning.core.html#pytorch_lightning.core.LightningModule.training_step) state that you can return a -1 to stop the current loop. Could this be useful in your case?

**Comment by david-alexander-white** [comment]
Ah ha! I think you've lead me to the solution here, although now I think the docs are wrong.

Relevant code
```
            # ---------------
            # RUN TRAIN STEP
            # ---------------
            output = self.run_training_batch(batch, batch_idx)
            batch_result, grad_norm_dic, batch_step_metrics = output

            # when returning -1 from train_step, we end epoch early
            early_stop_epoch = batch_result == -1
```

So we need a -1 in the first output of run_training_batch. I don't think returning -1 from training_step does that. However, when I was looking into this before I failed to notice that you CAN force it to return -1 by returning -1 from on_batch_start

so I think that's the solution -- to keep returning -1 from on_batch_start (may have to be a bunch of times depending on how many epochs you have)

**Comment by david-alexander-white** [comment]
> so I think that's the solution -- to keep returning -1 from on_batch_start (may have to be a bunch of times depending on how many epochs you have)

can confirm this works. might be nice to put that in the docs for early stopping as a third alternative, if no other changes will be made?

**Comment by Borda** [comment]
As mentioned on slack, @david-alexander-white very welcome to send a PR with doc update :rabbit:

**Comment by justusschock** [comment]
Although I agree, that it would increase flexibility if there was such a flag that would be checked

**Comment by jeremyjordan** [comment]
@justusschock  what’s the benefit of that versus raising some sort of exception to stop training? right now if you `raise KeyboardInterrupt` it should gracefully exit training. we can always have a custom exception too

**Comment by justusschock** [comment]
@jeremyjordan I simply don't like to use exceptions for something, that is somewhat expected (like a graceful shutdown or some kind of early stopping). IMO exceptions are just to handle errors correctly...

**Comment by jeremyjordan** [comment]
I added a Trainer attribute `should_stop` in #1504 which you can set to True to stop training after the current batch, once that PR is merged you should be able to use that from anywhere (callbacks, model hooks, etc).

**Comment by sachinruk** [comment]
@jeremyjordan just wondering if you could add this feature to the docs. Only found out about this from this question: https://github.com/PyTorchLightning/pytorch-lightning/discussions/7305#discussioncomment-682487

**Comment by nmichlo** [comment]
The current [Early Stopping](https://pytorch-lightning.readthedocs.io/en/latest/common/early_stopping.html) doc page does not list the `should_stop` attribute. I only found out about this here.

**Comment by LucaMarconato** [comment]
Setting `should_stop` exits the training phase, but the validation phase is still performed. How do I tell the trainer to completely end the training without further check of the validation loaders? `raise KeyboardInterrupt` is an hacky option for me because I am wrapping the training around `optuna`, so I would like to keep `raise KeyboardInterrupt` working as intended.

**Comment by tchaton** [comment]
Hey @ LucaMarconato,

It would be simply enable by doing the following there

```py
if not self.trainer.enable_validation or self.trainer.should_stop:
    return False
```

: https://github.com/PyTorchLightning/pytorch-lightning/blob/381343a79c703f2ccf1ab7c1d87400ad6e31fdf4/pytorch_lightning/loops/epoch/training_epoch_loop.py#L348

Would you be interested into making a contribution ?

Best,
T.C

**Comment by LucaMarconato** [comment]
Probably I am doing something wrong, but for me `self.trainer.should_stop = True` does not end training immediately and makes me have to deal with downstream exceptions. Like, if I have NaN in a tensor that makes some math operation nonsense, and I try to return a None from the loss function, I get an exception downstream from the early stopping. While if I return something like a torch.tensor([float('nan')]), another piece of the machinery complains that there is no gradient to do the backprop.

So I ended up raising a custom `RuntimeError` exception, and everything works as expected.

**Comment by LucaMarconato** [comment]
I have just noticed that throwing a custom exception leads to GPU memory leaks. So for me the problem is still open. Has anybody managed to use `self.trainer.should_stop` to terminate the training immediately? If the training does not stop immediately then I have to handle the cases in which the model continues to do a few training-related extra operations with garbage data (nans or inf).

**Comment by vedantroy** [comment]
Doing  `self.trainer.should_stop` is no good because it still runs the validation batch, which I don't want. I just want to immediately exit training. What would be the right thing to do here?

**Comment by CAIQT** [comment]
> Doing `self.trainer.should_stop` is no good because it still runs the validation batch, which I don't want. I just want to immediately exit training. What would be the right thing to do here?

also set:  self.trainer.limit_val_batches = 0  
may solve the problem.

**Comment by Toeplitz** [comment]
> > Doing `self.trainer.should_stop` is no good because it still runs the validation batch, which I don't want. I just want to immediately exit training. What would be the right thing to do here?
> 
> also set: self.trainer.limit_val_batches = 0 may solve the problem.

this worked for me
